README.md
# Jalaali plugin for Flutter [![pub package](https://img.shields.io/badge/pub-v0.4.0-blue.svg)](https://github.com/mrdimaan/flutter_jalaali)

A flutter package to make working with solar date easier. You can convert dates to solar dates and manipulate those by using this package in your app.

## Usage
First add `jalaali` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).
- [Documentation](https://mrdimaan.github.io/flutter_jalaali/)
- [Pub Home Page](https://pub.dartlang.org/packages/jalaali/)


CHANGELOG.md
## [0.4.0] - 4th January 2019
* Added `parse` constructor. see [documents](https://mrdimaan.github.io/flutter_jalaali/)

## [0.3.0] - 30th October 2018
* Added `format` method. see [documents](https://mrdimaan.github.io/flutter_jalaali/)

## [0.2.0] - 29th October 2018
* Changed the algorithm for convertion
* Removed extra constructors
* Added `toDateTime` method. see [documents](https://mrdimaan.github.io/flutter_jalaali/)

## [0.1.0] - 28th October 2018
* Under Developement
* Added some docs


test/jalaali_test.dart
import 'package:test/test.dart';
import 'package:jalaali/jalaali.dart';

void main() {
  test('Current Shamsi Date', () {
    var currentDate = JalaaliDate.now();
    print(currentDate.format("DD MMM YY", persianNumbers: true));
    expect(1, 1);
  });

  test('Date Parser', () {
    print(JalaaliDate.parse("2013-01-19 03:14:07"));
    expect(JalaaliDate.parse("2013-01-19 03:14:07"), "1391/10/30");
  });
}


lib/shamsi.dart
import 'package:jalaali/shared.dart';

/// Number of days in a cycle
const int cycle_days = 1029983;

/// Number of years in a cycle
const int cycle_years = 2820;

/// 365 + leapRatio
const double year_length = 365.24219858156028368;

/// 475/01/01 AP, start of 2820 cycle
const int hijri_shamsi_epoch = 2121446;

/// 683.0 / 2820.0
const double leap_threshold = 0.24219858156028368;

/// Determines if shamsi year is leap or not
bool shamsiIsLeap(int year) {
  double result = (year + 2346) * leap_threshold;
  double frac = result - result.toInt();
  if (frac < leap_threshold) {
    return true;
  } else {
    return false;
  }
}

/// Returns shamsi days in month
int shamsiDaysInMonth(int month, int year) {
  if (month > 0 && month <= 12) {
    return month < 7 ? 31 : month < 12 || shamsiIsLeap(year) ? 30 : 29;
  }
  return 0;
}

/// Returns shamsi days in year
int _shamsiDaysInYear(int year) {
  return shamsiIsLeap(year) ? 366 : 365;
}

// Returns shamsi months in tyear
int shamsiMonthInYear(int year) {
  return 12;
}

/// Returns cycle number
int _cycle(int jdn) {
  int offset = jdn - hijri_shamsi_epoch;
  int cycleNo = offset ~/ cycle_days;
  if (offset < 0) {
    --cycleNo;
  }
  return cycleNo;
}

int shamsiIsValid(int year, int month, int day) {
  if (year < 0) ++year;
  if (day > 0 && day <= shamsiDaysInMonth(month, year)) {
    return 1;
  }
  return 0;
}

int cycleStart(int jdn) {
  int era = _cycle(jdn);
  int start = hijri_shamsi_epoch + era * cycle_days;
  return start;
}

/// Returns first day of year in a cycle
int fdoyc(int year, int cycleNo) {
  /// Day in cycle
  double dc = (year * year_length);
  int fdoyc = dc.floor();
  return hijri_shamsi_epoch + cycleNo * cycle_days + fdoyc;
}

int fdoy(int year) {
  int cycleNumber = year ~/ cycle_years;
  // First day of year in a cycle
  int fdoyc;
  // Day number in cycle
  double dc;
  year -= 475;
  if (year < 0) {
    cycleNumber--;
  }
  year -= (cycleNumber * cycle_years);
  dc = (year * year_length);
  fdoyc = dc.floor();
  return hijri_shamsi_epoch + cycleNumber * cycle_days + fdoyc;
}

/// Convert shamsi to julian day number
int shamsiToJdn(int year, int month, int day) {
  /* Adjust the offset of year 0 */
  int era = 0;
  int dy = 0;
  int yc = 0;
  int fd = 0;
  int i = 0;
  if (year < 0) {
    ++year;
  }
  era = (year - 475) ~/ cycle_years;
  if ((year - 475) < 0) {
    --era;
  }
  yc = (year - 475) - era * cycle_years;
  fd = fdoyc(yc, era);
  dy = 0;
  for (i = 1; i < month; ++i) {
    dy += shamsiDaysInMonth(i, year);
  }
  dy += day;
  return fd + dy - 1;
}

DateData jdnToShamsi(int jd) {
  int c = _cycle(jd);
  int yc = ((jd - cycleStart(jd)) ~/ year_length);
  int y = yc + 475 + c * 2820;
  int d = jd - fdoyc(yc, c) + 1;
  int m = 0;
  if (d > _shamsiDaysInYear(y)) {
    y++;
    d = 1;
  }
  if (y <= 0) {
    y--;
  }
  for (m = 1; m < 12; ++m) {
    if (d > shamsiDaysInMonth(m, y)) {
      d -= shamsiDaysInMonth(m, y);
    } else {
      break;
    }
  }

  return new DateData(y, m, d);
}


lib/jalaali.dart
library jalaali;

import 'package:jalaali/gregorian.dart';
import 'package:jalaali/shamsi.dart';
import 'package:jalaali/shared.dart';
import 'package:jalaali/translator.dart';

class JalaaliDate {
  /// The day of the month [1..31]
  int day;

  /// The jalaali month [1..12]
  int month;

  /// The jalaali year
  int year;

  /// Temp instance of [DateTime] used for inital convertion value
  DateTime _dateTime;

  /// Default constructor from Jalali date
  JalaaliDate(int year, int month, int day)
      : year = year,
        month = month,
        day = day;

  /// Constructs a [JalaaliDate] instance with current date
  JalaaliDate.now() {
    _dateTime = new DateTime.now();
    _convertToShamsi(_dateTime.year, _dateTime.month, _dateTime.day);
  }

  /// Constructs a [JalaaliDate] instance from a datetime
  JalaaliDate.fromDateTime(DateTime datetime) : _dateTime = datetime {
    _convertToShamsi(_dateTime.year, _dateTime.month, _dateTime.day);
  }

  /// Constructs a [JalaaliDate] instance from parsing a date string
  JalaaliDate.parse(String stringDate) {
    _dateTime = DateTime.parse(stringDate);
    _convertToShamsi(_dateTime.year, _dateTime.month, _dateTime.day);
  }

  /// Returns a human-readable string for this instance.
  String toString({
    bool showDate = true,
    bool showTime = false,
  }) {
    DateTime.now().toString();
    if (!(showDate || showTime)) {
      throw new Exception(
          'At least one of arguments [showDate or showTime] must be true');
    }
    String stringDate = '';
    if (showDate) {
      stringDate =
          year.toString() + '/' + month.toString() + '/' + day.toString();
    }
    if (showTime) {
      stringDate += ' ' +
          _dateTime.hour.toString() +
          ':' +
          _dateTime.minute.toString() +
          ':' +
          _dateTime.second.toString();
    }
    return stringDate;
  }

  /// Returns a gregorian [DateData] from current shamsi date
  DateData _toGregorian() {
    int jdn = shamsiToJdn(year, month, day);
    DateData gregorin = jdnToGregorian(jdn);
    return gregorin;
  }

  /// Returns an instance of [DateTime] from this [JalaaliDate]
  DateTime toDateTime() {
    DateData date = _toGregorian();
    return new DateTime(date.year, date.month, date.day);
  }

  /// Converts gregorian date to jalaali date and stores to this instance of [JalaaliDate]
  void _convertToShamsi(int gyear, int gmonth, int gday) {
    int jdn = gregorianToJdn(gyear, gmonth, gday);
    DateData shamsi = jdnToShamsi(jdn);
    year = shamsi.year;
    month = shamsi.month;
    day = shamsi.day;
  }

  // Returns the jalaali date in determined format
  String format(String format, {bool persianNumbers = false}) {
    DateData jDate = DateData(year, month, day);
    return translate(format, jDate, persianNumbers);
  }
}


lib/gregorian.dart
import 'package:jalaali/math.dart';
import 'package:jalaali/shared.dart';

bool gregorianIsLeap(int year) {
  if ((year & 3) == 0 && ((year % 25) != 0 || (year & 15) == 0)) {
    return true;
  } else {
    return false;
  }
}

int gregorianDaysInMonth(int month, int year) {
  switch (month) {
    case 1:
    case 3:
    case 5:
    case 7:
    case 8:
    case 10:
    case 12:
      return 31;
      break;
    case 4:
    case 6:
    case 9:
    case 11:
      return 30;
      break;
    case 02:
      return gregorianIsLeap(year) ? 29 : 28;
      break;
    default:
      return 0;
      break;
  }
}

int gregorianDaysInYear(int year) {
  return gregorianIsLeap(year) ? 366 : 365;
}

int _gregorianMonthInYear(int year) {
  return 12;
}

bool gregorianIsValid(int year, int month, int day) {
  if (month <= _gregorianMonthInYear(year) &&
      day <= gregorianDaysInMonth(month, year)) {
    return true;
  }
  return false;
}

int gregorianToJdn(int year, int month, int day) {
  int c0 = fdiv((month - 3), 12);
  int x1 = month - (12 * c0) - 3;
  int x4 = year + c0;
  Div d = pdiv(x4, 100);
  int jd = fdiv(146097 * d.quot, 4) +
      fdiv(36525 * d.rem, 100) +
      fdiv(153 * x1 + 2, 5) +
      day +
      1721119;
  return jd;
}

DateData jdnToGregorian(int jd) {
  Div x3 = pdiv(4 * jd - 6884477, 146097);
  Div x2 = pdiv(100 * (x3.rem ~/ 4) + 99, 36525);
  Div x1 = pdiv(5 * (x2.rem ~/ 100) + 2, 153);
  int c0 = (x1.quot + 2) ~/ 12;

  int day = (x1.rem ~/ 5) + 1;
  int month = x1.quot - 12 * c0 + 3;
  int year = 100 * x3.quot + x2.quot + c0;

  return DateData(year, month, day);
}


lib/translator.dart
import 'package:jalaali/shared.dart';

/// Translates input format to a readable string date
String translate(String input, DateData date, bool persianNumbers) {
  input = input.replaceAll("YYYY", date.year.toString());
  input = input.replaceAll("YY", date.year.toString().substring(2, 4));
  input = input.replaceAll("MMM", _getJalaaliMonthName(date.month));
  input = input.replaceAll("MM", date.month.toString().padLeft(2, "0"));
  input = input.replaceAll("M", date.month.toString());
  input = input.replaceAll("DDD", _getPersianNumber(date.day));
  input = input.replaceAll("DD", date.day.toString().padLeft(2, "0"));
  input = input.replaceAll("D", date.day.toString());
  if (persianNumbers) {
    input = translateNumbers(input);
  }
  return input;
}

String translateNumbers(String input) {
  input = input.replaceAll("0", "۰");
  input = input.replaceAll("1", "۱");
  input = input.replaceAll("2", "۲");
  input = input.replaceAll("3", "۳");
  input = input.replaceAll("4", "۴");
  input = input.replaceAll("5", "۵");
  input = input.replaceAll("6", "۶");
  input = input.replaceAll("7", "۷");
  input = input.replaceAll("8", "۸");
  input = input.replaceAll("9", "۹");
  return input;
}

/// Returns jalaali month name
String _getJalaaliMonthName(int month) {
  switch (month) {
    case 1:
      return "فروردین";
    case 2:
      return "اردیبهشت";
    case 3:
      return "خرداد";
    case 4:
      return "تیر";
    case 5:
      return "مرداد";
    case 6:
      return "شهریور";
    case 7:
      return "مهر";
    case 8:
      return "آبان";
    case 9:
      return "آذر";
    case 10:
      return "دی";
    case 11:
      return "بهمن";
    case 12:
      return "اسفند";
    default:
      return "خطا";
  }
}

String _getPersianNumber(int number) {
  String output;
  switch (number) {
    case 1:
      output = "یک";
      break;
    case 2:
      output = "دو";
      break;
    case 3:
      output = "سه";
      break;
    case 4:
      output = "چهار";
      break;
    case 5:
      output = "پنج";
      break;
    case 6:
      output = "شش";
      break;
    case 7:
      output = "هفت";
      break;
    case 8:
      output = "هشت";
      break;
    case 9:
      output = "نه";
      break;
    case 10:
      output = "ده";
      break;
    case 11:
      output = "یازده";
      break;
    case 12:
      output = "دوازده";
      break;
    case 13:
      output = "سیزده";
      break;
    case 14:
      output = "چهارده";
      break;
    case 15:
      output = "پانزده";
      break;
    case 16:
      output = "شانزده";
      break;
    case 17:
      output = "هفده";
      break;
    case 18:
      output = "هجده";
      break;
    case 19:
      output = "نونزده";
      break;
    case 20:
      output = "بیست";
      break;
    case 21:
      output = "بیست و یک";
      break;
    case 22:
      output = "بیست و دو";
      break;
    case 23:
      output = "بیست و سه";
      break;
    case 24:
      output = "بیست و چهار";
      break;
    case 25:
      output = "بیست و پنج";
      break;
    case 26:
      output = "بیست و شش";
      break;
    case 27:
      output = "بیست و هفت";
      break;
    case 28:
      output = "بیست و هشت";
      break;
    case 29:
      output = "بیست و نه";
      break;
    case 30:
      output = "سی";
      break;
    case 31:
      output = "سی و یک";
      break;
    case 32:
      output = "سی و دو";
      break;
    default:
      return "خطا";
  }
  return output + "م";
}
